Fixing "Request Feedback" Behaviour

momo3404 opened the pull request:
Fixes #1035  .

Changes proposed in this PR:
- Changed the default feedback message.
- Modified "Sample Message" in the Organization Details -> Request Feedback section to match the editable feedback message.
- Added verification to ensure inappropriate placeholder data values are not saved with feedback message.
- Added explanation for placeholder data values.

![image](https://github.com/user-attachments/assets/89d63d50-7aa7-40a5-867d-a3ce75d92593)

- Edited the flash notice message that's shown after a user clicks "Request Feedback" when creating a plan
![image](https://github.com/user-attachments/assets/836d6b45-4945-4be5-a6b9-fb4edce81266)

momo3404 committed:
Edited Sample Message to Display Editable Feedback Message

momo3404 committed:
Flash notice shows correct message now

momo3404 committed:
Correct feedback message in plan

momo3404 committed:
Added new default feedback message

momo3404 committed:
made rubocop happy

momo3404 committed:
Edited Changelog

aaronskiba commented on app/views/plans/_request_feedback_form.html.erb:
```diff
@@ -12,7 +12,7 @@
        %>
        </p>
       <div class="well well-sm">
-        <%= sanitize plan.owner.org.feedback_msg.to_s % { user_name: current_user.name(false), plan_name: plan.title, organisation_email: current_user.org.contact_email } %>
```
- I think we need to continue using `plan.owner.org.feedback_msg` here.

momo3404 replied:
`plan.owner.org.feeback_msg` unfortunately leaves a bug where the wrong feedback message is displayed in the Plan -> Request Feedback section (in the grey box), changing it to `plan.org.feedback_msg` resolves that issue and displays the correct message for each organization

aaronskiba replied:
Sorry, is the bug the undesired feedback message?

aaronskiba replied:
I'm just thinking about the result for `plan.owner.org.feedback_msg` vs `plan.org.feedback_msg`.

`plan.owner.org` corresponds to the org of the user that created the plan.
`plan.org` corresponds to the org chosen when the plan was created.

This PR improves org.feedback_msg. But for now we are only applying the improved message to newly created orgs.

So I'm thinking there are two options to access the improved feedback message:
1) Your current change (`plan.owner.org.feedback_msg` -> `plan.org.feedback_msg`) when `plan.org` is a new org.
2) The existing implementation (keeping `plan.owner.org.feedback_msg`) when `user.org` is a new org.

If 2) is true, then I'm kind of leaning toward it. It would still enable the improved feedback message for users belonging to new orgs going forward.

Maintaining the current logic also makes sense to me when I think about things like `org.contact_email`.  Currently, that is matching the org that the plan owner is affiliated with.

momo3404 replied:
Yes, the bug is caused because the `plan.owner.org.feedback_msg` incorrectly displays the wrong feedback message for all organizations. This bug occurs because, for some reason, the first new organization created to test this fix is having its feedback message be displayed across all organizations in the plan section when `plan.owner.org.feedback_msg` is referenced.

On the other hand, `plan.org.feedback_msg` correctly displays the appropriate feedback message for each organization in the plan section, so I went with that. Let me know if you think it's still worth changing.

aaronskiba commented on app/views/orgs/_feedback_form.html.erb:
```diff
@@ -29,7 +29,11 @@
       </div>
       <div class="form-group col-xs-4">
         <h4><%= _('Sample Message') %></h4>
-        <%= sanitize sample_message_for_org_feedback_form(org) %>
```
- I'm encountering `TypeError: no implicit conversion of nil into String` for orgs where `org.contact_email` is nil. I think we'll have to add some extra handling for this case.
  - `OrgsHelper#sample_message_for_org_feedback_form` performs this handling. Maybe we should still be calling that method within this form?

momo3404 replied:
I think it's easier to just copy the email handling logic to the new line with the feedback message. I just pushed that and that seems to have solved the problem

aaronskiba replied:
Glad that solves the error.

This .gsub() chaining seems to work as well, but maybe we should stick to using format() or % for interpolating placeholders?

aaronskiba replied:
I'm also still kinda leaning toward keeping the view lean and still calling the existing `sample_message_for_org_feedback_form` helper method in `app/helpers/orgs_helper.rb`. I like that you added the accounting for the `%{plan_name}` placeholder here as well as the `(org.feedback_msg || '')` logic. The cleanest approach might be to just apply those changes to that helper method (i.e. Replace the hardcoded string with your `(org.feedback_msg || '')` logic and add a dummy value for `%{plan_name}`).

momo3404 committed:
Added fix for orgs with nil org email

momo3404 committed:
Returned plan feedback to correct behaviour

momo3404 committed:
Merge branch 'integration' into momo/issues/1035

aaronskiba commented on app/helpers/feedbacks_helper.rb:
```diff
@@ -7,11 +7,12 @@ def feedback_confirmation_default_subject
-      administrator at your organisation. " \
-      "If you have questions pertaining to this action, please contact us
-      at %{organisation_email}.</p>")
+    # This is the default feedback message
```
This is good. I'm just thinking that we'll probably want to apply these changes upstream at some point as well.
Maybe we could make the following updates in preparation for that?:
- Use `Rails.configuration.x.application.name` instead of "DMP Assistant"
- Use `Rails.configuration.x.organisation.helpdesk_email` rather than "dmp-assistant@tech.alliancecan.ca"

aaronskiba committed:
Add custom validator for feedback_msg interpolation

Adds a custom validator for the feedback message input on the `/org/admin/:id/admin_edit#feedback` path. Users can include placeholders in this message, which get interpolated with actual values on the `plans/:id/request_feedback` path.

This validator prevents the saving of problematic feedback_msg data by attempting to format the custom message with the sample `placeholders` hash and capturing any errors.

momo3404 committed:
added placeholder help msg

aaronskiba committed:
Update CHANGELOG.md

momo3404 committed:
Merge pull request #1068 from portagenetwork/aaron/momo/issues/1035

Add custom validator for `Org.feedback_msg` interpolation

momo3404 committed:
Merge branch 'momo/issues/1035' of https://github.com/portagenetwork/roadmap into momo/issues/1035

momo3404 committed:
Edited helper function to display feedback message

aaronskiba commented on app/views/plans/_request_feedback_form.html.erb:
```diff
@@ -12,7 +12,7 @@
        </p>
       <div class="well well-sm">
-        <%= sanitize plan.owner.org.feedback_msg.to_s % { user_name: current_user.name(false), plan_name: plan.title, organisation_email: current_user.org.contact_email } %>
+        <%= sanitize display_feedback_message(plan.owner.org, current_user, plan.owner.org.feedback_msg) %>
```
Nice, I like how app/views/orgs/_feedback_form.html.erb and app/views/plans/_request_feedback_form.html.erb are both using the helper method.

momo3404 committed:
Can display plan name in plan feedback message

aaronskiba commented on app/helpers/orgs_helper.rb:
```diff
@@ -3,17 +3,29 @@
+  rescue KeyError, ArgumentError => e
+    Rails.logger.error("Unable to display feedback message: #{e.message}")
+    message = _('Unable to render feedback message: ')
+    message + _(e.message.to_s)
```
Cool, I like that the plan_name is able to be displayed on the request feedback tab of the plan page now.

I'm thinking about one more potential thing here.
Pic 1 (below) I think this is useful for org admins as it enables them to see how to correct their feedback message.
![Screenshot from 2025-06-10 12-36-42](https://github.com/user-attachments/assets/24c31b7c-d169-46cc-ab5d-6b93a38308c0)

Pic 2 (below) The same error might be confusing for users that may not have any knowledge about the custom feedback message, placeholders, etc.
![Screenshot from 2025-06-10 12-38-27](https://github.com/user-attachments/assets/20e7c003-f54c-4cab-9d72-665b3bd93132)

If we were to change the request feedback tab on the plans page, I think we could just add some extra handling.

Maybe something like the following:
```ruby
  rescue KeyError, ArgumentError => e
    Rails.logger.error("Unable to display feedback message: #{e.message}")
    # /plans/:plan_id/request_feedback and /org/admin/:org_id/admin_edit both render the feedback_msg
    # If plan_name.nil?, then we are on /org/admin/:org_id/admin_edit
    if plan_name.nil?
      # Only render the error on the org admin page.
      message = _('Unable to render feedback message: ')
      message + _(e.message.to_s)
    end
  end
```
That should bypass the message altogether on the request feedback tab on the plan page. Do you think that would be an improvement?

momo3404 replied:
Good catch! Yea that's definitely better, I'll go ahead and push that

momo3404 committed:
Added placeholder explanation to org tab

momo3404 committed:
improved error handling with plan name

momo3404 committed:
made rubocop happy

momo3404 committed:
moved message label to correct spot

aaronskiba commented:
Manual testing checklist

- [x] Default feedback_msg is as expected for new orgs:
```
irb(main):001:0> Org.last.feedback_msg
  Org Load (0.6ms)  SELECT "orgs".* FROM "orgs" ORDER BY "orgs"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> "<p>Hello %{user_name},</p>A member of your organisation will respond to your request to review your data management plan\n       within 48 hours. If you have questions pertaining to this action please contact your local team at %{organisation_email}, or for assistance with DMP Assistant contact dmp-assistant@tech.alliancecan.ca."
```
![Screenshot from 2025-06-10 15-02-23](https://github.com/user-attachments/assets/c3547104-8f59-4948-822e-9d00b816c365)
- [x] Validator prevents saving of problematic feedback_msg
(When providing `%()` in feedback_msg)
![Screenshot from 2025-06-10 15-06-00](https://github.com/user-attachments/assets/8fd29ae3-2733-4728-b792-e660c5fffad7)
(When providing `%{bad_key}` in feedback_msg)
![Screenshot from 2025-06-10 15-05-20](https://github.com/user-attachments/assets/7856f554-808a-431f-b13d-2c68ec66af01)
- [x] Pre-existing problematic feedback_msg is rendered safely
![Screenshot from 2025-06-10 15-09-21](https://github.com/user-attachments/assets/336e5f95-4c52-44fe-b796-92d43db54162)
![Screenshot from 2025-06-10 15-09-37](https://github.com/user-attachments/assets/0e9ea446-0597-48d2-ac73-45825ec2375a)
- [x] Correct request feedback flash message is rendered
![Screenshot from 2025-06-10 15-10-16](https://github.com/user-attachments/assets/70aeb893-3289-4651-8a03-c6d838d8768c)
- [x] Sample message corresponds with org admin's inputted message
![Screenshot from 2025-06-10 15-14-07](https://github.com/user-attachments/assets/6a86a951-addf-4c02-ae96-da4e31b86a99)

aaronskiba commented on app/views/orgs/_feedback_form.html.erb:
```diff
@@ -23,13 +23,22 @@
+            <li><code style="color: #8A6E00;">%{organisation_email}</code> — <%= _('The contact email for your organisation') %></li>
+          </ul>
+          <strong><%= _('NOTE: Placeholders must be typed exactly as shown, including the %{...} format.') %></strong>
+        </p>
```
This all looks great, just one refactoring suggestion as a final, final change.

Our `*.scss` files definitely need some cleanup (hopefully we can tackle a bit of that as part of the bootstrap upgrade). But maybe you could just add something like the following to `app/assets/stylesheets/grey-branding.scss`:

```scss
.code-darkest-yellow {
  color: $color-alliance-darkest-yellow;
}
```
And then just change the code to use `<li><code class="code-darkest-yellow">`.

I'd suggest something similar for `<p class="help-block" style="color: #32322F;">`, but I think that could even be just replaced with `<p>` and the styling doesn't really seem to change.

momo3404 committed:
Refactored CSS in help message

momo3404 committed:
Merge branch 'integration' into momo/issues/1035

aaronskiba reviewed (approved):
Looks good. This one was a workout. Love the extra exception handling we did on the client and server sides. The placeholder explanation on the orgs page is a great addition too.
